ygen/enumgen.go

```diff
@@ -744,23 +747,36 @@ func (s *enumGenState) resolveNameClashSet(nameClashSets map[string]map[string]*
 	return uniqueNamesMap, nil
 }
 
-// resolveEnumeratedUnionEntry takes an input YANG union yang.Entry and computes the enumerated
-// values that should be generated for the entry. The name of the enumerated
-// value is calculated based on the original context, whether path compression is enabled based
-// on the compressPaths boolean, and whether the name should not include underscores, as per the
-// noUnderscores boolean.
+// resolveEnumeratedUnionEntry takes an input YANG union yang.Entry and
+// computes the names of the enumerated values that should be generated for the
+// entry. The name of each enumerated value is calculated based on the original
+// context: whether path compression is enabled, whether the name should not
+// include underscores, whether the enum name should be de-deduped if the same
+// enumeration is already been seen in a different part of the YANG tree,
+// whether a shortened version of the enumeration names should be given for
+// enumeration leaves, and whether the defining module name (as opposed to the
+// user module's name) should be used to name enumerated typedefs.
 func (s *enumGenState) resolveEnumeratedUnionEntry(e *yang.Entry, compressPaths, noUnderscores, skipEnumDedup, shortenEnumLeafNames, useDefiningModuleForTypedefEnumNames bool) error {
-	for _, t := range util.EnumeratedUnionTypes(e.Type.Type) {
+	enumTypes := util.EnumeratedUnionTypes(e.Type.Type)
+	subsumingTypes := util.EnumeratedUnionSubsumingTypes(e.Type)
+	if len(enumTypes) == 0 {
+		return fmt.Errorf("enumerated type had an empty union within it, path: %v, type: %v, enumerated: %v", e.Path(), e.Type, enumTypes)
+	}
+	for i, t := range enumTypes {
 		switch {
 		case t.IdentityBase != nil:
 			if err := s.resolveIdentityrefBaseTypeFromIdentity(t.IdentityBase, noUnderscores); err != nil {
 				return err
 			}
 		case t.Enum != nil:
-			if util.IsYANGBaseType(t) {
+			enumNameSake := t
+			if useDefiningModuleForTypedefEnumNames {
+				enumNameSake = subsumingTypes[i]
```

Comment:
To be clear here - `enumNameSake` is a type that is going to be used to give this type its name, because enumerations themselves are anonymous? I think it'd be useful to make a comment here that says something of this nature, just so that we remember this is how it works.